Add sudo and execute proposal types

ethanfrey opened the pull request:
Closes #672 
Closes #711

- [x] Define protobuf definitions
- [x] Wire up as proposals in wasm module
- [x] Implement proposal handlers
- [x] Test proposal handlers
- [x] Add cli commands for them
- [x] Add rest handlers for them
- [x] Wire up in cli/rest
- [x] Test cli tooling (manually?)

ethanfrey commented:
@alpe I think this is done and covered with tests. Can you please review and let me know how to test the cli better?

alpe commented on x/wasm/keeper/proposal_handler.go:
```diff
@@ -115,6 +119,53 @@ func handleMigrateProposal(ctx sdk.Context, k types.ContractOpsKeeper, p types.M
+
+	ctx.EventManager().EmitEvent(sdk.NewEvent(
+		types.EventTypeGovContractResult,
+		sdk.NewAttribute(types.AttributeKeyResultDataHex, hex.EncodeToString(data)),
```
👍

alpe commented on x/wasm/keeper/proposal_handler.go:
```diff
@@ -115,6 +119,53 @@ func handleMigrateProposal(ctx sdk.Context, k types.ContractOpsKeeper, p types.M
+	if err != nil {
+		return sdkerrors.Wrap(err, "run as address")
+	}
+	// we currently don't support sending tokens as part of an execute proposal. should we? from which account?
```
Good question! For me it would make sense to support this from the community pool for example.

ethanfrey replied:
Okay, I will add it to the proposal type

alpe commented on x/wasm/keeper/proposal_integration_test.go:
```diff
@@ -206,6 +206,119 @@ func TestMigrateProposal(t *testing.T) {
+
+	// balance should be empty (proper release)
+	bal = bankKeeper.GetBalance(ctx, contractAddr, "denom")
+	require.Equal(t, bal.Amount, sdk.NewInt(0))
```
👍 good test

alpe reviewed (approved):
LGTM 🎸 
Please update the method doc in https://github.com/CosmWasm/wasmd/blob/master/x/wasm/keeper/keeper.go#L441
as it would not be correct anymore.

If you can add the new types to `Governance.md` then this would be 🚀 

Disclaimer:
I have checked that the CLI commands are listed in `wasmd tx gov submit-proposal -h` but not executed a proposal

ethanfrey commented on proto/cosmwasm/wasm/v1/proposal.proto:
```diff
@@ -56,6 +56,7 @@ message MigrateContractProposal {
   string title = 1;
   // Description is a human readable text
   string description = 2;
+  // FIXME: I think this is unused? Migrate has no sender
```
Done in https://github.com/CosmWasm/wasmd/pull/730/commits/0a96e26fbad8ff67cd130ad151c7f8e968c12d09

ethanfrey commented on proto/cosmwasm/wasm/v1/proposal.proto:
```diff
@@ -66,6 +67,33 @@ message MigrateContractProposal {
+  string run_as = 3;
+  // Contract is the address of the smart contract
+  string contract = 4;
+  // Msg json encoded message to be passed to the contract as sudo
```
done in https://github.com/CosmWasm/wasmd/pull/730/commits/0a96e26fbad8ff67cd130ad151c7f8e968c12d09

ethanfrey commented:
Thank you for the feedback. I have addressed all comments, and also added the funds field to the new execute contract proposal type. I also updated the changelog and this is ready to merge
